### PR TITLE
Feature/line chart extra lines

### PR DESCRIPTION
**Pull request by agustka**
Adding extra line data to line charts that contains horizontal and vertical lines.

### CODE REVIEW

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 1]
```diff
@@ -129,6 +131,9 @@ class LineChartBarData {
   /// to show dot spots upon the line chart
   final FlDotData dotData;
 
+  // to show line chart annotations such as average line and vertical dot lines
+  final ExtraLinesData extraLinesData;
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Here we don't need it, because ExtraLinesData as defined on LineChartData, we should use that to draw lines in the whole chart, and not related any LineChartBarData,
Please remove it, and for the below lines as I said we should put it on the BelowBarData, I will explain it on the next sections.

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 2]
```diff
@@ -198,4 +204,62 @@ class FlDotData {
     this.dotSize = 4.0,
     this.checkToShowDot = showAllDots,
   });
-}
\ No newline at end of file
+}
+
+class DashDefinition {
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Please remove it, we can add line style functionality line Dash and Dot effects in next phase, here we just add two feature (Extra Lines -> you use it for avg line, and BelowBar Lines-> you use it for below bar lines), then I will add LineStyle after this PR, don't worry about that.

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 3]
```diff
@@ -198,4 +204,62 @@ class FlDotData {
...
+  const LineChartLine({
+    Color color,
+    double strokeWidth,
+    this.dashDefinition,
+  }) : super(color: color, strokeWidth: strokeWidth);
+}
+
+class VerticalLine extends LineChartLine {
+  final double x;
+  final double endY;
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: we don't need endY because ExtraLines should be drawn from y = 0 to y = height, and the only x is dynamic ( for the below line we should apply another solution), prefer this ExtraLines just for avg line.

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 4]
```diff
@@ -198,4 +204,62 @@ class FlDotData {
...
+    this.endY,
+    DashDefinition dashDefinition,
+    Color color = Colors.black,
+    double strokeWidth = 1,
+  }) : super(color: color, strokeWidth: strokeWidth, dashDefinition: dashDefinition);
+}
+
+class HorizontalLine extends LineChartLine {
+  final double y;
+  final double endX;
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: also we don't need endX, (read previous comment)

File: lib/src/chart/line_chart/line_chart_painter.dart [hunk 5]
```diff
@@ -40,6 +45,8 @@ class LineChartPainter extends AxisChartPainter {
     });
 
     drawTitles(canvas, viewSize);
+
+    drawAnnotation(canvas, viewSize);
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_painter.dart]: dude, this is not annotation in a generic definition, look at the problem as a generic problem, we should name it extraLines to draw extra lines on the chart, then you can use it for your problem to draw avg line, this is not related to the FL Chart.

File: lib/src/chart/line_chart/line_chart_painter.dart [hunk 6]
```diff
@@ -326,4 +333,85 @@ class LineChartPainter extends AxisChartPainter {
...
+      }
+      return;
+    }
+  }
+
+  void _drawHorizontalLineLine(Canvas canvas, Size viewSize, HorizontalLine line) {
+    viewSize = getChartUsableDrawSize(viewSize);
+    extraLinesPaint.color = line.color;
+    extraLinesPaint.strokeWidth = line.strokeWidth;
+    final double x = line.endX == null ? viewSize.width : getPixelX(line.endX, viewSize);
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_painter.dart]: here we can equal x with the viewSize.width every time.
because this is not for draw below lines

File: lib/src/chart/line_chart/line_chart_painter.dart [hunk 7]
```diff
@@ -326,4 +333,85 @@ class LineChartPainter extends AxisChartPainter {
...
+    final double y = getPixelY(line.y, viewSize);
+    if (line.dashDefinition != null) {
+      _drawDashedLine(canvas, extraLinesPaint, 0, y, x, y, line.dashDefinition);
+    } else {
+      _drawSolidLine(canvas, extraLinesPaint, 0, y, x, y);
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_painter.dart]: remove this if else and determine the line as a solid line every time ( I will add Line Style after your PR in a separate phase), and you can draw the solid line here and you don't need to put it in a new function (because there is no if else anymore)

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 4 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_data.dart]: What if the user wants the line to go all the way to the top though and not necessarily end at the data point?

File: lib/src/chart/line_chart/line_chart_painter.dart (hunk 5 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_painter.dart]: Residual naming from the original work when I first started. Renaming.

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 1 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_data.dart]: Remove it but put it where..?

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 2 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_data.dart]: All right, although I will not be able to use fl_chart library without it.

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 3 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_data.dart]: Yeah but what if the user wants the line to end at a data point for example?

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 1 above)
Comment (imaNNeoFighT) [reply lib/src/chart/line_chart/line_chart_data.dart]: As I saw you defined it to the LineChartData

Comment (agustka) [reply lib/src/chart/line_chart/line_chart_data.dart]: Removed.

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 2 above)
Comment (imaNNeoFighT) [reply lib/src/chart/line_chart/line_chart_data.dart]: We can not add all features in a single PR, I said multiple times that I will add it after your pR, it should add in a separate phase, because I want to reuse it on other parts such as grid and borders, this is a public library and we should implement all features generic and customizable, (for example add LineStyle on other parts).
be patient

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 3 above)
Comment (imaNNeoFighT) [reply lib/src/chart/line_chart/line_chart_data.dart]: If we want to draw this to the specific data point then we should have start data point alongside it.
I said, use it just for full horizontal and vertical lines, and it can be used as a avg line in your usecase, then you should use BelowBarData to draw lines below dot points

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 4 above)
Comment (imaNNeoFighT) [reply lib/src/chart/line_chart/line_chart_data.dart]: Drop it plz, if we want to make it possible to set end point then we have to add start point for it to be customizable

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 3 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_data.dart]: After checking it some more I understand better what below bar data is for, so I'm dropping endY.

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 8]
```diff
@@ -162,6 +164,9 @@ class BelowBarData {
   final Offset gradientFrom;
   final Offset gradientTo;
 
+
+  final List<VerticalLine> verticalLines;
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: here we should just have a single FlLine that determines the style of line, then we can repeat it for all spot points, we don't need VerticalLine, because we know where we should draw them (below the spots)

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 9]
```diff
@@ -198,4 +204,37 @@ class FlDotData {
...
+
+  const VerticalLine({
+    @required this.x,
+    Color color = Colors.black,
+    double strokeWidth = 1,
+  }) : super(color: color, strokeWidth: strokeWidth);
+}
+
+class HorizontalLine extends FlLine {
+  final double y;
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: We have a convention on Library, 
the horizontal is all of the things that draw on the x axis,
the vertical is all of the things that draw on the y axis,
then HorizontalLine means lines that draw on the x axis and ( for example in x = 1, 2, 3, 4, 5), and y is the same  (may be it is VerticalLine in your head, but it is a convention in this library, check the FlTitlesData you can find out what does it means.)
conclusion:
we should have x here instead of y

File: lib/src/chart/line_chart/line_chart_painter.dart (hunk 6 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_painter.dart]: Removed

File: lib/src/chart/line_chart/line_chart_painter.dart (hunk 7 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_painter.dart]: Removed

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 8 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_data.dart]: Dropped in favor of FlLine

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 9 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_data.dart]: Won't this confuse everybody? A horizontal line has infinite x values... seems strange but I'll change, its your library.

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 4 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_data.dart]: Dropped

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 9 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_data.dart]: Y has been changed to x.

File: lib/src/chart/line_chart/line_chart_painter.dart [hunk 10]
```diff
@@ -326,4 +348,31 @@ class LineChartPainter extends AxisChartPainter {
 
   @override
   bool shouldRepaint(CustomPainter oldDelegate) => false;
+
+  void drawExtraLines(Canvas canvas, Size viewSize) {
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_painter.dart]: Here plz draw both horizontal and vertical lines, (firstly data class should be updated)

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 11]
```diff
@@ -198,4 +204,27 @@ class FlDotData {
...
+  final double x;
+
+  const HorizontalLine({
+    @required this.x,
+    Color color = Colors.black,
+    double strokeWidth = 1,
+  }) : super(color: color, strokeWidth: strokeWidth);
+}
+
+/// This class holds data about drawing chart annotations (data decorations) such as average line and data point lines
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Update the comment, ( we don't need to mention anything about annotation, this word is specific for your use case)

Comment (agustka) [reply lib/src/chart/line_chart/line_chart_data.dart]: Fixed

File: lib/src/chart/line_chart/line_chart_painter.dart (hunk 10 above)
Comment (agustka) [reply lib/src/chart/line_chart/line_chart_painter.dart]: The drawing of the two line types has been merged into this method.

Comment (agustka) [reply lib/src/chart/line_chart/line_chart_painter.dart]: But I'm not sure what you mean with that the data class should be updated?

File: lib/src/chart/line_chart/line_chart_painter.dart [hunk 12]
```diff
@@ -317,6 +323,36 @@ class LineChartPainter extends AxisChartPainter {
     }
   }
 
+  void drawExtraLines(Canvas canvas, Size viewSize) {
+    // Draw vertical data point lines under spots
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_painter.dart]: her is not under spots, fix the comment

File: lib/src/chart/line_chart/line_chart_painter.dart [hunk 13]
```diff
@@ -317,6 +323,36 @@ class LineChartPainter extends AxisChartPainter {
...
+    viewSize = getChartUsableDrawSize(viewSize);
+    for (LineChartBarData barData in data.lineBarsData) {
+      if (barData.belowBarData.verticalLines == null) {
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_painter.dart]: why you are reading it from belowBarData?
dude draw it just like the horizontal lines,
you should have a class like this:
```
class ExtraLinesData {
  final bool show;

   // Horizontal line definitions
  final List<HorizontalLine> horizontalLines;

  // Vertical line definitions
  final List<VerticalLine> verticalLines;

   const ExtraLinesData({
    this.show = true,
    this.horizontalLines,
    this.verticalLines,
  });
}
```

both of them should define in this class

Comment (agustka) [reply lib/src/chart/line_chart/line_chart_painter.dart]: You told me to move it to below bar data...
